### DISCUSSION TITLE
Communication between Python and Rust threads/asyncloops

### DISCUSSION THREAD

**Discussion by roma-glushko** [PyO3/pyo3#3438] [Questions] [answered]
Hey folks 👋 

I'm a new guy to the PyO3 and its community, but I have gotten an interesting case to implement and wanted to consult with you on where to start.

I want to implement a class in Rust that would either run in a separate thread or thread + event loop. When started, the class is going to do some work (until explicitly stopped) and will put the results somewhere (normally, it would be some sort of queue). 

At the same time, Python will be running in its own thread and will be pulling those results and processing them further.

So both parties (Python & Rust) will be running at their own paces and would be sharing just some sort of queue. 

What is the best way to archive this sort of evilibrium? I would appreciate any information, examples that would seed my further investigation. 

Thank you in advance 🙏

**Comment by davidhewitt** [comment, marked as answer]
Sorry for the delayed reply. Are you using `async`? If so, `pyo3-asyncio` may be able to help solve your problems and provide examples.

If you want to send Python and Rust jobs between threads, maybe https://github.com/PyO3/pyo3/issues/576#issuecomment-1713039140 will be interesting to you.

**Reply by roma-glushko** [reply]
@davidhewitt hey David 👋   thank you for the answer!

I'm not using `pyo3-asyncio`  yet, but heard about it from the pyo3 docs. Even though I want this project to be working with Python's asyncio, I would probably try to keep it as uniform as possible on the Rust side. With that respect, I like what these projects does to keep Rust part minimalistic (which is good for me as I'm a Rust novice):

- https://github.com/simmsb/pyo3_queue_test_eventfd/blob/master/src/lib.rs 
- https://github.com/samuelcolvin/watchfiles/blob/main/watchfiles/main.py (I like this one less as it does simple short-looping, I believe it could be implement much efficiently)

> If you want to send Python and Rust jobs between threads, maybe https://github.com/PyO3/pyo3/issues/576#issuecomment-1713039140 will be interesting to you.

This is quite a big thread 😄 

This reference seems like what might need: https://github.com/letalboy/RustPyNet/blob/main/RustPyNet/src/python_pool/pool.rs Although it does a lot more work with Python on the Rust end than I would imagine.

Ideally, what I want is a ~~magic~~ `std::sync::mpsc::channel()`-like channel that I could create and then pass its sender in a separate Rust thread while allowing Python's world to access the receiver via some method on the Python-Rust-bound class. The Rust thread would generate messages/data objects and put them to the channel and Python would check that channel out to read and further process the pending messages.

**Reply by davidhewitt** [reply]
Sounds like you have a good understanding of the solution. You could wrap the receiver end of the channel in a `#[pyclass]` to expose the receiving end to Python. You have the choice of using blocking or async to wait for messages.

**Reply by roma-glushko** [reply]
@davidhewitt I feel like I came really close to it but I have one issue that perplexes me.

I ended up using the `crossbeam-channels` lib to communicate between threads. They took the idea of Golang channels (that's another way to explain what I wanted to attain) and ported to Rust. Then I created a new thread scope and add the message generation logic there:

https://github.com/roma-glushko/inotify-toolkit/blob/c6316829098ca47fcfcf5ba6f1b1174901cbb010/src/watcher.rs#L406-L415

Although, the issue is that the whole thing gets hung up when the new thread is up. Seems like it does not really give any control back to the Python's main thread for some reason. Like it's trying to join the new thread immediately. 

Somewhat equivalent Rust program works well though:

```rs
use notify::{Config, RecommendedWatcher, RecursiveMode, Watcher};
use std::path::Path;
use std::thread;

use crossbeam_channel::{unbounded, Sender, Receiver};

fn main() {
    if let Err(e) = watch("./wd") {
        println!("{:?}", e)
    }
}

fn watch<P: AsRef<Path>>(path: P) -> notify::Result<()> {
    let (tx, rx) = unbounded();

    let mut watcher = RecommendedWatcher::new(tx, Config::default())?;
    watcher.watch(path.as_ref(), RecursiveMode::Recursive)?;

    thread::scope(|scope| {
        scope.spawn(move || {
            for res in rx {
                match res {
                    Ok(event) => println!("Change: {event:?}"),
                    Err(error) => println!("Error: {error:?}"),
                }
            }
        });
    });

    Ok(())
}
```

Have you seen this problem before, David?

**Reply by davidhewitt** [reply]
Are you potentially missing a call to `Python::allow_threads` around the point where the thread blocks to wait on the channel?

**Reply by roma-glushko** [reply]
@davidhewitt this one is not that easy to hop over. Have tried to all `allow_threads()` all around the places, but still could not fix it 🤔 

https://github.com/roma-glushko/inotify-toolkit/commit/228025c0a461f293bbe766939f4bdb0f1caac1e6

**Reply by davidhewitt** [reply]
Your choice of `crossbeam_thread::scope` is probably the cause. That requires the watcher thread to finish before it will ever return. Probably you just need to use `std::thread::spawn` (and save a reference to that thread if you need to kill it later).

You probably need to change your `__enter__` method to the signature:

```
pub fn __enter__(slf: Py<self>, py: Python<'_>) -> PyResult<()> {
```

so that you're able to move the `slf` smart pointer inside the thread.

**Reply by roma-glushko** [reply]
@davidhewitt thank you for the help David!

What I ended up doing was separating the main logic out of PyO3 wrapper: https://github.com/roma-glushko/notifykit/blob/main/src/watcher.rs#L53

That really helped me to get reference passing right and now the main logic works 🙌
